ipam/pool.go

```diff
@@ -512,20 +512,31 @@ func (ap *addressPool) requestAddress(address string, options map[string]string)
 }
 
 // Releases a previously requested address back to its address pool.
-func (ap *addressPool) releaseAddress(address string) error {
+func (ap *addressPool) releaseAddress(address string, options map[string]string) error {
 	var err error
 
 	log.Printf("[ipam] Releasing address %v.", address)
 	defer func() { log.Printf("[ipam] Address release completed with err:%v.", err) }()
 
 	ar := ap.Addresses[address]
 	if ar == nil {
+		id := options[OptAddressID]
 		// Handle pre-assigned addresses.
 		if address == ap.Gateway.String() {
 			return nil
+		} else if id != "" {
+			// Get address associated with id.
+			ar = ap.addrsByID[id]
+			if ar == nil {
+				err = errAddressNotFound
+				return err
+			}
+			address = ar.Addr.String()
+			log.Printf("[ipam] Releasing the address %v associated with id %v", address, id)
+		} else {
+			err = errAddressNotFound
```

Comment:
No need to repeat the address not found case. Move this to outside the if block.